app/assets/javascripts/govuk-component/option-select.js

```diff
@@ -99,16 +102,50 @@
     return this.$optionSelect.hasClass('js-closed');
   };
 
+  OptionSelect.prototype.setContainerHeight = function setContainerHeight(height){
+    this.$optionsContainer.css({
+      'max-height': 'none', // Have to clear the 'max-height' set by the CSS in order for 'height' to be applied
+      'height': height
+    });
+  };
+
+  OptionSelect.prototype.isLabelVisible = function isLabelVisible(index, option){
+    var $label = $(option);
+    var initialOptionContainerHeight = this.$optionsContainer.height();
+    var optionListOffsetTop = this.$optionList.offset().top;
+    var distanceFromTopOfContainer = $label.offset().top - optionListOffsetTop;
+    return distanceFromTopOfContainer < initialOptionContainerHeight;
+  };
+
+  OptionSelect.prototype.getVisibleLabels = function getVisibleLabels(){
+    return this.$labels.filter(this.isLabelVisible.bind(this));
+  };
+
   OptionSelect.prototype.setupHeight = function setupHeight(){
-    var optionsContainer = this.$optionSelect.find('.options-container');
-    var optionList = optionsContainer.children('.js-auto-height-inner');
-    var initialOptionContainerHeight = optionsContainer.height();
-    var height = optionList.height();
+    var initialOptionContainerHeight = this.$optionsContainer.height();
+    var height = this.$optionList.height();
+    var lastVisibleLabel;
 
     if (height < initialOptionContainerHeight + 50) {
       // Resize if the list is only slightly bigger than its container
-      optionsContainer.height(optionList.height());
+      this.setContainerHeight(height);
+      return;
     }
+
+    // Resize to cut last item cleanly in half
+    lastVisibleLabel = this.getVisibleLabels().last();
+
+    this.setContainerHeight(
+      lastVisibleLabel.position().top +
+      parseInt(lastVisibleLabel.css('border-top-width'), 10) +
+      parseInt(lastVisibleLabel.css('padding-top'), 10) +
+      (parseInt(
+        "normal" == lastVisibleLabel.css('line-height') ?
+          lastVisibleLabel.css('font-size') : lastVisibleLabel.css('line-height'),
+        10
+      ) / 2)
```

Comment:
This block is quite hard to interpret. It might be worthwhile splitting out the various parts, logic and parsing into named variables beforehand, so the arithmetic is understandable at a glance.

Something like `this.setContainerHeight(position + topBorder + topPadding + halfOfLabelHeight)`